mlir/lib/Dialect/Affine/Transforms/FoldMemRefAliasOps.cpp

```diff
@@ -0,0 +1,267 @@
+  LogicalResult matchAndRewrite(AffineStoreOp storeOp,
+                                PatternRewriter &rewriter) const override {
+    auto expandShapeOp =
+        storeOp.getMemref().getDefiningOp<memref::ExpandShapeOp>();
+
+    if (!expandShapeOp)
+      return failure();
+
+    SmallVector<Value> indices;
+    expandToUnderlyingIndices(storeOp.getAffineMap(), storeOp.getIndices(),
+                              storeOp.getLoc(), rewriter, indices);
+
+    SmallVector<Value> sourceIndices;
+    // affine.store guarantees that indexes start inbounds, which impacts if our
+    // linearization is `disjoint`.
+    if (failed(memref::resolveSourceIndicesExpandShape(
+            storeOp.getLoc(), rewriter, expandShapeOp, indices, sourceIndices,
+            /*startsInbounds=*/true)))
+      return failure();
+
+    rewriter.replaceOpWithNewOp<AffineStoreOp>(
+        storeOp, storeOp.getValueToStore(), expandShapeOp.getViewSource(),
+        sourceIndices);
+    return success();
+  }
+};
+
+class AffineStoreOpOfCollapseShapeOpFolder final
+    : public OpRewritePattern<AffineStoreOp> {
+public:
+  using OpRewritePattern<AffineStoreOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(AffineStoreOp storeOp,
+                                PatternRewriter &rewriter) const override {
+    auto collapseShapeOp =
+        storeOp.getMemref().getDefiningOp<memref::CollapseShapeOp>();
+
+    if (!collapseShapeOp)
+      return failure();
+
+    // For affine ops, we need to apply the map to get the "actual" indices.
+    SmallVector<Value> indices;
+    expandToUnderlyingIndices(storeOp.getAffineMap(), storeOp.getIndices(),
+                              storeOp.getLoc(), rewriter, indices);
+
+    SmallVector<Value> sourceIndices;
+    if (failed(memref::resolveSourceIndicesCollapseShape(
+            storeOp.getLoc(), rewriter, collapseShapeOp, indices,
+            sourceIndices)))
+      return failure();
+
+    rewriter.replaceOpWithNewOp<AffineStoreOp>(
+        storeOp, storeOp.getValueToStore(), collapseShapeOp.getViewSource(),
+        sourceIndices);
+    return success();
+  }
+};
+
+} // namespace
+
+void affine::populateAffineFoldMemRefAliasOpPatterns(
+    RewritePatternSet &patterns) {
+  patterns
+      .add<AffineLoadOpOfSubViewOpFolder, AffineLoadOpOfExpandShapeOpFolder,
+           AffineLoadOpOfCollapseShapeOpFolder, AffineStoreOpOfSubViewOpFolder,
+           AffineStoreOpOfExpandShapeOpFolder,
+           AffineStoreOpOfCollapseShapeOpFolder>(patterns.getContext());
+}
+
+//===----------------------------------------------------------------------===//
+// Pass registration
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+struct AffineFoldMemRefAliasOpsPass final
+    : public affine::impl::AffineFoldMemRefAliasOpsBase<
+          AffineFoldMemRefAliasOpsPass> {
+  void runOnOperation() override;
+};
+
+} // namespace
+
+void AffineFoldMemRefAliasOpsPass::runOnOperation() {
```

Comment:
nit: I don't think there is much sense is doing such small methods out of line.

Reply:
It feels like pretty common style in MLIR passes to have the declaration at the top and the `runOnOperation` all the way down at the bottom

Reply:
It feels like pretty common style in MLIR passes to have the declaration at the top and the `runOnOperation` all the way down at the bottom